### ISSUE TITLE
Exception in CUDA simulator before syncthreads deadlocks process exit.

### ISSUE THREAD

**Issue by asford** [numba/numba#3063]
An exception occurring *before* a call to `syncthreads` in the CUDA simulator properly reraises the source exception, however it leaves the remaining threads of the kernel invocation blocked on their `syncthreads` events. This then blocks `sys.exit`, as the process waits indefinitely on the event-blocked non-daemon kernel threads.

This is a reasonably high impact issue in cases where the CUDA simulator is used to provide simple test coverage where CUDA is not available (ie. many CI contexts). If a test case fails due to an exception within the simulator the test process will deadlock, rather than exit with a non-zero status.

The script below a minimal repro when run with numba 0.38. The process will print an `IndexError` and traceback, as expected, but will not exit. A traceback on `threading.join` will be printed after receiving `SIGINT`.

```python
import numpy
import numba
numba.config.ENABLE_CUDASIM = 1

import numba.cuda as cuda

@numba.cuda.jit
def assign_with_sync(x, y):
    i = cuda.grid(1)
    y[i] = x[i]

    cuda.syncthreads()

size = 16
assign_with_sync[1, size * 2](numpy.arange(size), numpy.empty(size))
```

**Comment by asford** [comment]
Possible solutions:

1. Spawn simulator threads as daemon threads during [thread launch](https://github.com/numba/numba/blob/0.38.1/numba/cuda/simulator/kernel.py#L125) allowing process exit with blocked threads. This is a minimal workaround, however it allows the simulator to "leak" blocked threads on kernel exceptions. This is non-ideal as, as mentioned above, this may cause issues in test contexts where many simulator kernels are launched.

2. Add a "abort" flag member to `BlockThread`, set the abort flag on and unblock *all* threads when [a kernel exception is reraised](https://github.com/numba/numba/blob/0.38.1/numba/cuda/simulator/kernel.py#L200), and check-and-except if the abort flag is set in the simulator [`syncthreads` handler](https://github.com/numba/numba/blob/0.38.1/numba/cuda/simulator/kernel.py#L147). This is obviously a slightly more complex solution, but has the benefit of properly closing all kernel threads if the kernel exits with an exception.

**Comment by stuartarchibald** [comment]
Thanks for the report, I can reproduce. Thanks also for the proposed fix in #3067  :)
